subs/pantry/src/Pantry.hs

```diff
@@ -1501,3 +1505,36 @@ prunePackageWithDeps pkgs getName getDeps (pname, a)  = do
       else do
         modify' $ first (Map.insert pname prunedDeps)
       return $ not (null prunedDeps)
+
```

Comment:
Possible alternative API for the three functions below:

```haskell
withSnapshotCache
  :: (HasPantryConfig env, HasLogFunc env)
  => SnapshotCacheHash
  -> RIO env (Map PackageName (Set ModuleName))
  -> ((ModuleName -> RIO env [PackageName]) -> RIO env a)
  -> RIO env a
```

Reasoning: you can provide this function an action which will generate the full snapshot cache in memory on demand. This function can be responsible for checking if the cache already exists and, if not, creating it. This function can also create a single SQLite connection and reuse it for the entire body, bypassing the need to restart a database transaction. Finally, it can internally use the `SnapshotCacheId` without leaking that information to the caller, possibly providing more efficiency.

Reply:
Yeah, I had somewhat similar idea while working on this but I didn't pursue it further somehow. Will do it now, thanks.

Reply:
Done in 4d82cddeb0a69d14ab16f26d7c023a8872f4c073 with one exception - SQLite connection is not shared, the main reason is that code reading cabal files also goes into Pantry and requests its own DB connection